Document how IC resolves host collisions (#1183)
nginxinc/kubernetes-ingress: docs-web/configuration/handling-host-collisions.md

# Handling Host Collisions

A host collision occurs when multiple resources configure the same `host`. The Ingress Controller supports two options for handling host collisions:
* Choosing the winner so that only one resource handles the host.
* Merging configuration of the conflicting resources.

## Choosing the Winner

Consider the following two resources:
* `cafe-ingress` Ingress:
    ```yaml
    apiVersion: networking.k8s.io/v1beta1
    kind: Ingress
    metadata:
      name: cafe-ingress
      annotations:
        kubernetes.io/ingress.class: "nginx"
    spec:
      rules:
      - host: cafe.example.com
        . . .
    ```
* `cafe-virtual-server` VirtualServer:
    ```yaml
    apiVersion: k8s.nginx.org/v1
    kind: VirtualServer
    metadata:
      name: cafe-virtual-server
    spec:

>       host: cafe.example.co

Reviewer: ```suggestion
      host: cafe.example.com
```

      . . .
    ```

If a user creates both resources in the cluster, a host collision will occur. As a result, the Ingress Controller will pick the winner using the following algorithm:

> > If multiple resources contend for the same host, the Ingress Controller will pick the winner based on the `creationTimestamp` of the resources: the oldest resource will win. In case there are more than one oldest resources (their `creationTimestamp` is the same),  the Ingress Controller will choose the resource with the smallest `uid`. Note: the `creationTimestamp` and `uid` fields are part of the resource [ObjectMeta](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#objectmeta-v1-meta). 

Reviewer 2: ```suggestion
> If multiple resources contend for the same host, the Ingress Controller will pick the winner based on the `creationTimestamp` of the resources: the oldest resource will win. If the `creationTimestamp` of both resources is the same, the Ingress Controller will choose the resource with the lower `uid`. 

> Note: The `creationTimestamp` and `uid` fields are part of the resource's [ObjectMeta](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#objectmeta-v1-meta). 
```

Author: > If the `creationTimestamp` of both resources...

in theory, more than 2 resources can have the same `creationTimestamp`, so that's why I didn't use `both` there.

> choose the resource with the lower `uid`

here are examples of `uid` - `dd50953e-3910-4d0a-8971-17a75b1aa1d5`, `bd6b403b-7f26-482c-8c06-23f11de24a73`.

 I don't think `lower` would apply here, because uids are not numbers like `5` and `6`.

I see how my original description `smaller` could be confusing as well. How about `lexicographically smaller`?  This reflects that to determine the winner in that case, we need to compare strings lexicographically.

> In our example, if `cafe-virtual-server` was created first, it will  win the host `cafe.example.com`, and the Ingress Controller will reject `cafe-ingress`. The winning and losing will be reflected in the events and the status field of the resources:

Reviewer 2: ```suggestion
In our example, if `cafe-virtual-server` was created first, it will win the host `cafe.example.com` and the Ingress Controller will reject `cafe-ingress`. This will be reflected in the events and in the resource's status field:
```

```
$ kubectl describe vs cafe-virtual-server
. . .
Status:
  . . .
  Message:  Configuration for default/cafe-virtual-server was added or updated
  Reason:   AddedOrUpdated
  State:    Valid
Events:
  Type    Reason          Age   From                      Message
  ----    ------          ----  ----                      -------
  Normal  AddedOrUpdated  9s    nginx-ingress-controller  Configuration for default/cafe-virtual-server was added or updated

$ kubectl describe ingress cafe-ingress
. . .
Events:
  Type     Reason    Age   From                      Message
  ----     ------    ----  ----                      -------
  Warning  Rejected  66s   nginx-ingress-controller  All hosts are taken by other resources
```

> > Note: Ingress resources allow configuring multiple hosts. As a result, an Ingress resource can be the winner for some of its hosts and a loser for the rest. For example, if `cafe-ingress` had an additional rule host rule `pub.example.com`, the Ingress Controller would not reject the Ingress: it would allow `cafe-ingress` to handle `pub.example.com`.

Reviewer 2: ```suggestion
> Note: You can configure multiple hosts for Ingress resources. As a result, it's possible that an Ingress resource can be the winner for some of its hosts and a loser for the others. For example, if `cafe-ingress` had an additional rule host rule -- `pub.example.com` -- the Ingress Controller would not reject the Ingress. Rather, it would allow `cafe-ingress` to handle `pub.example.com`.
```

> Similarly, if `cafe-ingress` was created first, it will win `cafe.example.com`, and the Ingress Controller will reject `cafe-virtual-server`.

Reviewer 2: ```suggestion
Similarly, if `cafe-ingress` was created first, it will win `cafe.example.com` and the Ingress Controller will reject `cafe-virtual-server`.
```

## Merging Configuration for the Same Host

> It is possible to merge configuration for multiple Ingress resources for the same host. One common use case for that is distributing such resources across multiple namespaces. See the [Cross-namespace Configuration](/nginx-ingress-controller/configuration/ingress-resources/cross-namespace-configuration/) doc.

Reviewer 2: ```suggestion
It is possible to merge configuration for multiple Ingress resources for the same host. One common use case for this approach is distributing resources across multiple namespaces. See the [Cross-namespace Configuration](/nginx-ingress-controller/configuration/ingress-resources/cross-namespace-configuration/) doc for more information.
```

> It is *not* possible to merge configuration for multiple VirtualServer resources for the same host. However, you can split such VirtualServers into multiple VirtualServerRoute resources that a single VirtualServer can then reference. See the [corresponding example](https://github.com/nginxinc/kubernetes-ingress/tree/master/examples-of-custom-resources/cross-namespace-configuration) on GitHub.

Reviewer 2: ```suggestion
It is *not* possible to merge the configurations for multiple VirtualServer resources for the same host. However, you can split the VirtualServers into multiple VirtualServerRoute resources, which a single VirtualServer can then reference. See the [corresponding example](https://github.com/nginxinc/kubernetes-ingress/tree/master/examples-of-custom-resources/cross-namespace-configuration) on GitHub.
```
